counting: add counting v2 command (#334)
yagpdb-cc/yagpdb-cc: website/docs/fun/counting_v2.md

---
title: Counting System
---

> This command controls the counting game and the statistics command.

Reviewer: A brief comment about what additional features your command offers over the previous counting command would be nice.

## Trigger

**Type:** `Regex`<br />

> **Trigger:** `.+`

Reviewer: Note that `.+` does not match text split over multiple lines. I have not examined your code thoroughly but it seems like this could be a concern; for instance, would
```
1
<random text here>
```
be marked as invalid, assuming `1` is the next number?

Note also that this does not trigger on image-only posts, which might be abused? Not sure.

Author: Your example is valid and correct given that the number or expression on the first line evaluates to the next number.
I'm unsure how image-only posts could be used to abuse the system. My intent in many of my commands has been to allow "outside conversation" within the channel while not disrupting the game at hand. To do this, but also allow for such a wide range of input, I chose this trigger.

Reviewer: Fair enough. That is a significant difference from most counting systems I have seen in the wild, so I think it is worth mentioning explicitly in the description of this command (that unrelated content is permitted.)

Author: As messages that do include text content will trigger the command and an output if all characters do not match the available numerals, operators, or functions, maybe my claim for "outside conversation" is invalid for this command.
For example, if a user were to post "1+1 equals 2", this would be taken as invalid and a message would be output to show that.
So while messages without text will be ignored/permitted, messages with text while maybe unrelated, will not be entirely ignored. They will not affect the system or modify internal values.
So where I am now is how to word this in the description, or if it's even worth including.

Reviewer: We should certainly document that the command does not reject invalid inputs by deleting messages, since many other counting systems do so and users might mistakenly assume that this command does the same. Including additional detail beyond that is your choice; I do not have a strong opinion on documenting that.

:::note Restrictions

Set this command to _only run_ in your counting channel in the channel restrictions.

:::

## Usage

- Type a number, roman numeral, or a mix of both... with the option of math incrementing by 1, starting with 1, in the counting channel!
- `-CountingStatistics` - Show server stats for counting
- `-CountingStatistics <user>` - Show user stats for counting
- `-CountingStatistics Leaderboard` - Show server leaderboard for number of times counted correct by each user

:::tip Math Functionality

To view all available math functions, see [Supported Functions, Operators, and Constants](https://github.com/ei14/calc#supported-functions-operators-and-constants)

:::

## Configuration

- `$countTwice`<br />
  Whether to allow users to count multiple times in a row.

- `$correctRID`<br />
  ID of role for most recent correct counting user.

- `$incorrectRID`<br />
  ID of role for incorrect counting users.

- `$errorCID`<br />
  ID of channel for any error output.

- `$saves`<br />

>   Amount of saves from incorrect counts before the count resets.

Reviewer: It's not clear what "saves from incorrect counts" means. Elaborate?

Author: The amount of times an incorrect number or expression can be made before the count resets.
If $saves is 1, the first incorrect number sent will get a pass, but the second will reset the count to 1.

Reviewer: Thanks, but I meant that you should clarify the documentation.

---
title: Counting System
---

> This command controls the counting game and the statistics command.

Reviewer: ```suggestion
This command implements a counting game with detailed statistics and a leaderboard.
```

> This version of a counting command allows for the usage of base 10 and roman numerals, optionally joined by the usage of math functions and/or operators to count from 1, incrementing by 1, until failure.

Reviewer: ```suggestion
A unique feature of this command is that, besides decimal numbers, it also accepts roman numerals and mathematical expressions as valid inputs. So if the next number is 2, `II` and `2*sin(pi/2)` would be considered correct as well.
```
I've not checked whether `2*sin(pi/2)` actually works with your code; replace with an appropriate example if not.

Author: Just checked, it does. The built-in calc command handles the calculation, the CC just converts values into an acceptable expression and invalidates if necessary.

> Also included are statistics for the server and each user for how many times they have counted and how many times they've been correct.

Reviewer: You can drop this sentence after changing the first line of the documentation as suggested earlier.

> Messages without text are ignored by the system, effectively allowing unrelated content. However, messages including text which cannot be translated into a math expression will be quoted as invalid, but not deleted.

Reviewer: I think this should be in a `::note` callout for visibility. Concerning the content, I suggest
```suggestion
Unlike some other counting systems, this command does not delete invalid numbers to allow counting and other conversation to take place concurrently. Messages that include text which do not correspond to valid numbers will, however, trigger a warning message.
```

A clarifying question: is it the case that completely unrelated messages that do not contain anything resembling a number or a mathematical expression (so something like `Hello there!`) trigger the invalid input warning as well? If so, I would think this would disrupt conversation to the extent that you might as well disallow invalid inputs completely and delete them as most counting games do. But if you feel strongly that this command is fine as is, I am happy to drop this point.

Author: Yes, messages such as `Hello there!` will trigger the warning message. I agree with you though, and will work on allowing messages as such without warning for them.

Author: Messed around with the expression regex and trying to find logic that would determine the message is not meant to be an expression. Issue is the regex, due to the part which captures roman numerals, also captures empty strings, which I'd prefer it not but cannot figure out how to require at least one roman numeral (I, V, X, L, C, D, M) for it to be captured as a roman numeral. Another issue is, using the logic `if eq (joinStr "" $re) ""`, if the message includes the common letter e, or another character related to math, it evaluates to false, therefore disqualifying that method.

Author: Update: As of the latest commit - which makes a change to the regex which parses the expression and adds a conditional statement to make sure an expression was parsed - I believe this command to act fully as intended. 

Messages which are not meant to be expressions are to be ignored, message which are attempted expressions but have invalid syntax or include unaccepted characters/"functions" are to be given a warning message, and valid expressions are to be accepted and evaluated towards the correct next number.

---
title: Counting System
---

This command implements a counting game with detailed statistics and a leaderboard.

A unique feature of this command is that, besides decimal numbers, it also accepts roman numerals and mathematical expressions as valid inputs. So if the next number is 2, `II` and `2*sin(pi/2)` would be considered correct as well.

:::note

> Unlike some other counting systems, this command does not delete invalid numbers to allow counting and other conversation to take place concurrently. Messages that include text which do not correspond to valid numbers will, however, trigger a warning message.

Author: I think your first suggestion reads somewhat incorrectly; it reads to me that if I were to throw a 1 somewhere in the message, it would be valid, even if the expression sent is `invalid1exp`. However, this does make me notice the usage of the word "numbers" here, which may make the user question if terms other than decimal numbers are accepted. I think it'd be better to use "terms" in replacement.

As for the question and suggestion on first line expressions, yes, only the first line of the message is parsed for an expression. I believe the previous paragraph was the result of a [previous discussion](https://github.com/yagpdb-cc/yagpdb-cc/pull/334#discussion_r1685176559). It could use some expanding to make it known only the first line is parsed and the rest is ignored.

I will not make any changes to this now so it can be discussed.

> Unlike some other counting systems, this command does not delete invalid numbers to allow counting and other conversation to take place concurrently. Messages that include text which do not correspond to valid numbers will, however, trigger a warning message.

Author: ```suggestion
Unlike some other counting systems, this command does not delete invalid expressions to allow counting and other conversation to take place concurrently. Expressions which include unaccepted terms or evaluate to an invalid expression will, however, trigger a warning message.
```

> Unlike some other counting systems, this command does not delete invalid numbers to allow counting and other conversation to take place concurrently. Messages that include text which do not correspond to valid numbers will, however, trigger a warning message.

Reviewer 2: > I think your first suggestion reads somewhat incorrectly; it reads to me that if I were to throw a 1 somewhere in the message, it would be valid, even if the expression sent is invalid1exp

That was not the point of the first suggestion. The point was to simplify this sentence, the second suggestion expanded on that by asking to note down the behaviour with regards to the first line of a message.

> I think it'd be better to use "terms" in replacement.

No, you already defined what counts as a valid number for the purposes of this custom command, why introduce another term? There's "number" and "valid number", I don't see how replacing that with "expression" makes things clearer.

Even still, your suggestion doesn't make this any clearer anyway, as you're not stating that this "expression" has to occur on the first line.

> Unlike some other counting systems, this command does not delete invalid numbers to allow counting and other conversation to take place concurrently. Messages that include text which do not correspond to valid numbers will, however, trigger a warning message.

Author: I understand the point of the first suggestion was for simplification. However, I am saying that from the rewrite, it does not read the same.

I think using "number" in the docs will not influence the users to use the command to it's potential, which is fine if they don't want to use roman numerals and/or math. But I think it scopes in on decimal numbers as the message to send, and not the result of an expression.

I did not include the second line as I wasn't confident in including it, and can be committed separately after this discussion. Overall, yes, I agree it should be noted only the first line is to be parsed.

> Unlike some other counting systems, this command does not delete invalid numbers to allow counting and other conversation to take place concurrently. Messages that include text which do not correspond to valid numbers will, however, trigger a warning message.

Reviewer 2: Sounds like we came to an agreement. Add these notes and then we should be good to go.

> Unlike some other counting systems, this command does not delete invalid numbers to allow counting and other conversation to take place concurrently. Messages that include text which do not correspond to valid numbers will, however, trigger a warning message.

Author: I've updated the note to be more direct in my opinion and added the first line suggestion.
